### PR TITLE
Add NonPrivateXCTestMembersRule

### PR DESCRIPTION
**Pull request by keith**
This rule lints against any members of XCTestCase subclasses that aren't
implementing XCTestCase methods, or prefixed with `test`. This can help
encourage reducing ACLs on these classes, and potentially find
mis-spelled test functions.

### CODE REVIEW

**Comment by keith** [comment]
Open questions:

- [ ] Should we provide a way to identify custom test subclasses so this validation handles those as well (potentially similar to https://github.com/realm/SwiftLint/pull/3192)
- [ ] Should we explicitly ignore `override`s? Probably but probably only matters if we implement the above
- [ ] Instead of linting against ACL > private / fileprivate maybe instead we should only lint against not explicitly defined ACLs, since if you mark it as public and that's not validated somewhere else maybe that will always be by design (potentially this difference in behavior could be behind a setting)

**Comment by keith** [comment]
Still relevant

**Comment by jpsim** [comment]
Are you still working through the open questions? Would you like my take on them?

**Comment by keith** [comment]
I'd love your take!

File: Source/SwiftLintFramework/Rules/Lint/NonPrivateXCTestMembersRule.swift
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/Lint/NonPrivateXCTestMembersRule.swift:35]: This has the potential of making code that previously compiled stop compiling in several cases:

1. If there was an explicit ACL keyword used
2. If the type was used either outside the file (internal) or outside the type scope (fileprivate)

I'm comfortable with those tradeoffs but let me know if you're not.

**Comment by jpsim** [comment]
> Should we provide a way to identify custom test subclasses so this validation handles those as well

I'm comfortable keeping things simple to start and saying no to this one. This is an opt-in rule and we can always add support for this if there's an appetite.

> Should we explicitly ignore overrides? Probably but probably only matters if we implement the above

Again, I'm ok keeping things simple and not handling this right now.

> Instead of linting against ACL > private / fileprivate maybe instead we should only lint against not explicitly defined ACLs, since if you mark it as public and that's not validated somewhere else maybe that will always be by design (potentially this difference in behavior could be behind a setting)

I think you see where I'm going with this. Let's start with keeping things simple.

In most cases it's probably better to use composition for test helpers rather than subclassing anyway IMHO.

**Review by jpsim — APPROVED** [review]
👍 This looks good to me. @keith I made some small changes (made the rule correctable, turned it on in SwiftLint) so I'll let you merge this if/when you're comfortable with its state.

**Pull request merged**
